test/monitoring/components_metrics_test.go

```diff
@@ -92,19 +88,49 @@ func testComponentsPrometheusMetrics(t *testing.T, v1api v1.API) {
 
 			const contextTimeout = 10
 
-			ctx, cancel := context.WithTimeout(context.Background(), contextTimeout*time.Second)
-			defer cancel()
+			var err error
 
-			results, warnings, err := v1api.Query(ctx, tc.query, time.Now())
-			if err != nil {
-				t.Errorf("error querying Prometheus: %v", err)
-				return
+			// This loop ensures that we try to query prometheus for the state of targets multiple times.
+			// This is the retry logic.
+			for i := 0; i < 20; i++ {
```

Comment:
I was wondering if the nested retry logic helps? 

So here we are trying (1) one metric 20 times vs (2) trying all metrics 20 times?

With the first approach the logic is not being fair with all the tests. Which means if the test cases that fail because the data was not reconciled in the prometheus will incur other tests for not being tested.

With 2nd logic we give all the tests enough time to reconcile and in the last run we can know who hasn't reconciled yet.

Reply:
Hm, I think we should add `t.Parallel()` then, so each metric is tested in parallel, not sequentially. Then the retry logic will make sense, right?

Reply:
I think just inversing the loops should help the retry loop outside and the testcases inside.

Reply:
I disagree. Why would I again test metrics, which I already know they are fine?

Reply:
> Which means if the test cases that fail because the data was not reconciled in the prometheus will incur other tests for not being tested.

Also, I don't think this is true. All sub-tests are executed independently, so other metrics will still be evaluated.

Reply:
Oh just saw the other comments load, to figure out which tests are passed and need skipping maybe you can add a flag if they are passed or not in the struct and you can just skip them next time. But I still think that the enclosing should be like this 

`retry loop -> testcases `

and not `testscases -> retry looop`.

Reply:
I don't understand it. What are your arguments to favor this way of looping instead of existing one?

With my approach, all cases are tested in parallel and only until they succeed, which usually happens at first attempt.

With your approach, I either need to implement custom logic to skip the tests which already passed or test passing cases multiple times, despite the fact that they already passed.